### PR TITLE
Add dimensional counters for internal metrics

### PR DESCRIPTION
**Pull request by kokje**
This is a POC to enable finer granularity in emission tracking by adding dimensional counters of dpm instead of one single counter "MetricsSent". Since this is a first pass, there is definitely more testing to be done. Some of my comments/concerns are:
- the map needs a lock during reads/writes because maps can grow dynamically. Should I use a fixed length array instead? It might be a little arbitrary to determine what a good length is..
- counters are incremented in a separate goroutine to maintain separation from existing functionality  (also because I think incrementing internal counters is a concurrent and not parallel process wrt to our problem statement).  Due to this we lose ordering and it is possible that a read might occur before a write in which case there would be a delay in seeing the most accurate numbers. However these are cumulative counters and the windows are pretty small (~5s - 10s) so it may be worth the trade off

### CODE REVIEW

**Comment by poros** [comment]
Why don't we add a dimension to the "fullerite.collector_datapoints" metric instead? I think that this logic shouldn't be dependent on the handler (and so on the metrics storage system), but on the collector logic instead.

File: src/fullerite/handler/handler.go [hunk 1]
```diff
@@ -536,7 +552,21 @@ func (base *BaseHandler) emitAndTime(
...
 	} else {
 		atomic.AddUint64(&base.metricsDropped, uint64(numMetrics))
 	}
 }
+
+func (base *BaseHandler) countByDimension(
+	metrics []metric.Metric,
+) {
+	mu.Lock()
+	for i := range metrics {
```
Comment (tiras-j) [inline src/fullerite/handler/handler.go]: Tiny thing, but you can avoid indexing into the list by iterating with the value using `for i, item := range metrics`.  Just a bit more idiomatic.

File: src/fullerite/handler/handler.go [hunk 2]
```diff
@@ -342,6 +350,14 @@ func (base *BaseHandler) InternalMetrics() metric.InternalMetrics {
 		"metricsDropped": float64(base.metricsDropped),
 		"metricsSent":    float64(base.metricsSent),
 	}
+
+	if base.metricDimension != "" {
```
Comment (tiras-j) [inline src/fullerite/handler/handler.go]: Is this a config flag on whether to emit meta-metrics broken down by dimension? If so the name should be a bit clearer, on first reading it seemed you were assigning a dimension to the base handler or something.

Comment (tiras-j) [inline src/fullerite/handler/handler.go]: Ah nevermind, I see this is the dimension on which you parameterize your counting. That makes more sense.

File: src/fullerite/handler/handler.go [hunk 3]
```diff
@@ -536,7 +552,21 @@ func (base *BaseHandler) emitAndTime(
...
+		go base.countByDimension(metrics)
 	} else {
 		atomic.AddUint64(&base.metricsDropped, uint64(numMetrics))
 	}
 }
+
+func (base *BaseHandler) countByDimension(
+	metrics []metric.Metric,
+) {
+	mu.Lock()
```
Comment (gnufied) [inline src/fullerite/handler/handler.go]: prefer `defer mu.Unlock()`

**Comment by dichiarafrancesco** [comment]
@kokje Don't you think we should do this grouping by collector? What if two collectors are using the same dimension name with two different meanings? 

Also I'm concerned to the performance of this change, do we want to inspect all the metrics we send? Can we find a way to do the same thing but when the metrics are created? I like @poros suggestion for this.

**Comment by kokje** [comment]
Thanks all for the feedback! I will try moving the counters to the collector section but it may not be as straightforward as having them in the handler..especially given that we have blacklisted collectors. Meanwhile we're also testing/profiling changes from #324

**Comment by kokje** [comment]
Added dimensional counters to collectors, for now I've left both options in but leaning towards this being in calculated in the collectors.
@dichiarafrancesco++ good point, having the same dimension that means different things between collectors might be an issue. Also wrt performance I'm not too concerned about inspecting each metric because goroutines are pretty fast and number of emissions in the window are relatively small

File: src/fullerite/collectors.go [hunk 4]
```diff
@@ -108,6 +109,13 @@ func readFromCollector(collector collector.Collector,
 			continue
 		}
 		emissionCounter[c]++
+		if attributionDimension := collector.MetricsAttributionDimension(); attributionDimension != "" {
+			if val, ok := m.GetDimensionValue(attributionDimension); ok {
+				metricName := c + "_by_" + strings.Replace(val, " ", "_", -1)
```
Comment (poros) [inline src/fullerite/collectors.go]: Can't we just use a dimension here instead of generating a new metric?

Comment (kokje) [inline src/fullerite/collectors.go]: This gets attached as a dimension on fullerite.collector_datapoints, I guess I should rename the variable

Comment (poros) [inline src/fullerite/collectors.go]: Would you mind giving me an example of how the metric would look like, then?

**Comment by dichiarafrancesco** [comment]
A part from the name thing that I didn't quite understand this code lg2m!

**Pull request closed without merging**